docs/src/examples/nlp_tricks.jl

```diff
@@ -0,0 +1,124 @@
+# Copyright 2017, Iain Dunning, Joey Huchette, Miles Lubin, and contributors    #src
+# This Source Code Form is subject to the terms of the Mozilla Public License   #src
+# v.2.0. If a copy of the MPL was not distributed with this file, You can       #src
+# obtain one at https://mozilla.org/MPL/2.0/.                                   #src
+
+# # Nonlinear tips and tricks
+
+# This example collates some tips and tricks you can use when formulating
+# nonlinear programs. It uses the following packages:
+
+using JuMP
+import Ipopt
+import Test
+
+# ## User-defined functions with vector outputs
+
+# A common situation is to have a user-defined function like the following that
+# returns multiple outputs (we define `function_calls` to keep track of how
+# many times we call this method):
+
+function_calls = 0
+function foo(x, y)
+    global function_calls += 1
+    common_term = x^2 + y^2
+    term_1 = sqrt(1 + common_term)
+    term_2 = common_term
+    return term_1, term_2
+end
+
+# For example, the first term might be used in the objective, and the second
+# term might be used in a constraint, and often they share share work that is
+# expensive to evaluate.
+
+# This is a problem for JuMP, because it requires user-defined functions to
+# return a single number. One option is to define two separate functions, the
+# first returning the first argument, and the second returning the second
+# argument.
+
+foo_1(x, y) = foo(x, y)[1]
+foo_2(x, y) = foo(x, y)[2]
+
+# However, if the common term is expensive to compute, this approach is wasteful
+# because it will evaluate the expensive term twice. Let's have a look at how
+# many times we evalute `x^2 + y^2` during a solve:
+
+model = Model(Ipopt.Optimizer)
+set_silent(model)
+@variable(model, x[1:2] >= 0, start = 0.1)
+register(model, :foo_1, 2, foo_1; autodiff = true)
+register(model, :foo_2, 2, foo_2; autodiff = true)
+@NLobjective(model, Max, foo_1(x[1], x[2]))
+@NLconstraint(model, foo_2(x[1], x[2]) <= 2)
+function_calls = 0
+optimize!(model)
+Test.@test objective_value(model) ≈ √3 atol=1e-4
+Test.@test value.(x) ≈ [1.0, 1.0] atol=1e-4
+println("Naive approach: function calls = $(function_calls)")
+
+# An alternative approach is to use _memoization_, which uses a cache to store
+# the result of function evaluations. We can write a memoization function as
+# follows:
+
+"""
+    memoize(foo::Function, n_outputs::Int)
+
+Take a function `foo` and return a vector of length `n_outputs`, where each
+element is a function that returns the `i`'th output of `foo`. To avoid
+duplication of work, cache the evaluations of `foo`.
+"""
+function memoize(foo::Function, n_outputs::Int)
+    cache = Dict{UInt, Any}()
```

Comment:
I think it's cleaner and a bit more memory efficient to store only the last evaluation point `last_x` and the value `foo(last_x)`. Then, check if x is equal to `last_x` and use `foo(last_x)` if it is.

Reply:
The difficulty is Ipopt doesn't call `foo_1` and `foo_2` sequentially, so checking `last_x` doesn't work. (You really need to check the last ~6, and that adds an equal amount of complexity). I've added a `max_cache_size` to deal with the memory issue.

Reply:
Why do you need to check the last 6?

Reply:
Hmm. When I set the `max_cache_size` to less than that it was using more than 20 evaluations. But you're right, it only needs to cache the last `Float64` evaluation and the last `ForwardDiff.Dual` evaluation.